src/components/TodoList.jsx

```diff
@@ -1,64 +1,151 @@
         };
-
         this.addTask = this.addTask.bind(this);
         this.removeTask = this.removeTask.bind(this);
         this.onInputChange = this.onInputChange.bind(this);
+        this.removeAllTasks = this.removeAllTasks.bind(this)
+        this.sortTasks = this.sortTasks.bind(this)
+        this.setTaskInEditText = this.setTaskInEditText.bind(this)
+        this.setNewTaskText = this.setNewTaskText.bind(this)
     }
-
-    removeTask(text) {
+    // Set text of task in edit mode 
+    setTaskInEditText(taskText) {
+        this.setState({
+            ...this.state,
+            taskInEditText: taskText
+        })
+    }
+    // Change name of task
+    setNewTaskText(text, newText) {
+        const { tasks } = this.state
+        const id = this.getTaskNames().indexOf(text)
+        tasks[id].taskName = newText
+        this.setState({
+            ...this.state,
+            tasks
+        })
+    }
+    // Update state with sorted tasks
+    sortTasks(event) {
         const { tasks } = this.state;
+        const option = event.target.value
+        this.setState({
+            ...this.state,
+            tasks: tasks.sort(filterTasks(option)),
+            sortingOption: option
+        });
+    }
 
+    // Array of names of tasks
+    getTaskNames() {
+        const taskNames = [];
+        for (const task of this.state.tasks) {
+            taskNames.push(task.taskName)
+        }
+        return taskNames
+    }
 
+    removeTask(todoPointName) {
+        const { tasks } = this.state;
+        tasks.splice(this.getTaskNames().indexOf(todoPointName), 1)
         const newState = {
-            ...this.state,
-            tasks: tasks.filter(elem => elem != text)
-        }
+            ...this.state
+        };
         this.setState(newState);
     }
 
     addTask() {
-        const { tasks } = this.state;
-
-        this.setState({
+        const { tasks, sortingOption: sortingOpt } = this.state;
+        const newTask = {
+            taskName: this.state.input,
+            date: new Date()
+        };
+        // Sort task as it added
+        const newTasks = tasks.concat(newTask).sort(filterTasks(sortingOpt));
+        const newState = {
             input: "",
-            tasks: tasks.concat(this.state.input)
-        });
+            tasks: newTasks
+        }
+        this.setState(newState);
     }
 
     onInputChange(event) {
         const text = event.target.value;
-        this.setState({ ...this.state, input: text });
+        this.setState({ ...this.state, input: text })
+    }
+
+    removeAllTasks() {
+        this.setState(
+            {
+                ...this.state,
+                tasks: []
+            }
+        )
+    }
+
+    mapTasks(mapArgument) {
+        return this.state.tasks.map(mapArgument)
+    }
+    // Key of task html element 
+    getKeyForTodoElem(todoPoint) {
+        return this.state.tasks.indexOf(todoPoint)
     }
 
     render() {
         return <div>
```

Comment:
You see this extra div, wrapping around all the elements? React requests that any JSX return is inside a single element, so we can't just throw away this div. However, we can use a JSX "sugar" called **fragments**.

If you need to return JSX, but don't wanna create extra HTML nested elements, you can use fragments like this:
```jsx
<>
// Your elements here
</>
```

So it's an empty HTML element, basically. Note, that when React renders this - it won't create any HTML elements in place of the fragment, so you are saving on HTML nesting by using them, which is good for performance and later page analysis. In your case, you can do:

![image](https://user-images.githubusercontent.com/3951613/89110118-3a5f0780-d450-11ea-9ef4-c9c0b880019a.png)